### ISSUE TITLE
fix python warnings in generated notebooks

### ISSUE THREAD

**Issue by fkazemian** [QuantEcon/lecture-source-py#9]
![7](https://user-images.githubusercontent.com/20714542/44705480-16ffe500-aae2-11e8-9947-b83ac831745c.png)

**Comment by mmcky** [comment]
thanks @fkazemian this is due to the generated notebook having been executed. It looks like this lecture needs to be updated to remove the warning. @natashawatkins can you look into this issue? @fkazemian can you please add lecture reference.

**Comment by natashawatkins** [comment]
@mmcky should we be hiding warnings? There are quite a lot of warnings that show up in notebooks that we have generally ignored ie. divide by 0 warnings, invalid log inputs (when applying an operation over an array)

**Comment by mmcky** [comment]
@natashawatkins my inclination is to address these warnings as they are typically meaningful for the code that it is executing. I think many of them should be fixed by changing the code. I don't think we should hide warnings as they will reduce the changes of picking them and changing them.

**Comment by mmcky** [comment]
@fkazemian would you mind documenting which notebooks currently contain `Warnings` in output blocks?

**Comment by fkazemian** [comment]
Hi @mmcky 
I checked warnings from `.ipynb` files:
`grep -r -i --include=\*.rst 'Warning' ./`
```
./source/py/markov_asset.rst:    warning_message = "Spectral radius condition fails"
./source/py/markov_asset.rst:    assert np.max(np.abs(eigvals(K))) < 1 / β,  warning_message
./source/py/amss3.rst:**Warning:** Key equations in  :cite:`BEGS1` section III.D carry  typos  that we correct below
./source/py/mle.rst:Here we received a warning message saying "Maximum number of iterations has been exceeded."
```

and also for checking from `html` files:
```
./source/py/markov_asset.rst:    warning_message = "Spectral radius condition fails"
./source/py/markov_asset.rst:    assert np.max(np.abs(eigvals(K))) < 1 / β,  warning_message
./source/py/amss3.rst:**Warning:** Key equations in  :cite:`BEGS1` section III.D carry  typos  that we correct below
./source/py/mle.rst:Here we received a warning message saying "Maximum number of iterations has been exceeded."

```

**Comment by natashawatkins** [comment]
@fkazemian that doesn't seem to be a complete list - ie. the initial example isn't there

**Comment by fkazemian** [comment]
Hi @natashawatkins , Thanks, so I'll go through all of the lectures from compare-lectures.quantecon.org (latest build) and document the errors and warnings from outputs here.

**Comment by fkazemian** [comment]
- [ ] py_source/about_py.rst 
- [ ] py_source/additive_functionals.rst
- [ ] py_source/aiyagari.rst
- [ ] py_source/amss.rst
- [ ] py_source/amss2.rst
- [ ] py_source/amss3.rst
- [ ] py_source/arellano.rst
- [ ] py_source/arma.rst
- [ ] py_source/calvo.rst
- [ ] py_source/career.rst
- [ ] py_source/chang_credible.rst
- [ ] py_source/chang_ramsey.rst
- [ ] py_source/classical_filtering.rst
- [ ] py_source/coleman_policy_iter.rst

- additive_functionals
![screenshot from 2018-09-19 10-25-23](https://user-images.githubusercontent.com/20714542/45723818-77bca200-bbf6-11e8-9ab1-49346e70f658.png)

- amss     There is no warning, but until block 5 we have outputs but after that we can see the lack of `output`s
![screenshot from 2018-09-19 10-33-20](https://user-images.githubusercontent.com/20714542/45724135-04b42b00-bbf8-11e8-93d1-21be2d9c317d.png)

![screenshot from 2018-09-19 10-33-12](https://user-images.githubusercontent.com/20714542/45724136-04b42b00-bbf8-11e8-9464-32e2776d0c30.png)

![screenshot from 2018-09-19 10-32-58](https://user-images.githubusercontent.com/20714542/45724137-04b42b00-bbf8-11e8-84fa-54d5e944a9a3.png)

![screenshot from 2018-09-19 10-32-44](https://user-images.githubusercontent.com/20714542/45724138-04b42b00-bbf8-11e8-966d-890cbf0b5153.png)

-  amss2
![screenshot from 2018-09-19 10-39-15](https://user-images.githubusercontent.com/20714542/45724217-6d9ba300-bbf8-11e8-8d4d-3b2e4aa33a51.png)

-  amss3
![screenshot from 2018-09-19 10-52-24](https://user-images.githubusercontent.com/20714542/45724607-4c3bb680-bbfa-11e8-9028-ab48b82777a5.png)

-  arma
![screenshot from 2018-09-19 10-57-21](https://user-images.githubusercontent.com/20714542/45724756-f1568f00-bbfa-11e8-8bb6-645d01d65eda.png)

-  chang_credible
![screenshot from 2018-09-19 11-04-56](https://user-images.githubusercontent.com/20714542/45725004-04b62a00-bbfc-11e8-9cad-2219f615b486.png)
 
and also lack of the output for last two blocks
![screenshot from 2018-09-19 11-05-27](https://user-images.githubusercontent.com/20714542/45725022-1992bd80-bbfc-11e8-868d-54b234effbd0.png)

-  chang_ramsey
![screenshot from 2018-09-19 11-18-07](https://user-images.githubusercontent.com/20714542/45725346-cf124080-bbfd-11e8-9484-4ab6535869df.png)
and after that, we can see lack of the output for all blocks
![screenshot from 2018-09-19 11-18-18](https://user-images.githubusercontent.com/20714542/45725378-fc5eee80-bbfd-11e8-8004-dedc1bdab654.png)

**Comment by fkazemian** [comment]
- [x] - py_source/debugging.rst
- [ ] - py_source/discrete_dp.rst
- [x] - py_source/dyn_stack.rst
- [x] - py_source/egm_policy_iter.rst
- [ ] - py_source/estspec.rst
- [x] - py_source/finite_markov.rst
- [x] - py_source/getting_started.rst
- [x] - py_source/harrison_kreps.rst
- [x] - py_source/hist_dep_policies.rst
- [x] - py_source/ifp.rst
- [x] - py_source/jv.rst
- [x] - py_source/kalman.rst


- discrete_dp
![screenshot from 2018-09-19 11-39-13](https://user-images.githubusercontent.com/20714542/45726038-eef73380-bc00-11e8-819c-c7e1fa7d11f9.png)

- estspec

![screenshot from 2018-09-19 11-44-41](https://user-images.githubusercontent.com/20714542/45726176-94120c00-bc01-11e8-815d-53949ff4e240.png)

![screenshot from 2018-09-19 11-44-30](https://user-images.githubusercontent.com/20714542/45726177-94120c00-bc01-11e8-9eea-abd149f0e787.png)

![screenshot from 2018-09-19 11-43-56](https://user-images.githubusercontent.com/20714542/45726178-94120c00-bc01-11e8-9794-db342dad133c.png)

- finite_markov
![screenshot from 2018-09-19 11-48-32](https://user-images.githubusercontent.com/20714542/45726326-08e54600-bc02-11e8-99fd-a1533c443ce1.png)

- hist_dep_policies 
twice output 
![screenshot from 2018-09-19 11-52-32](https://user-images.githubusercontent.com/20714542/45726454-afc9e200-bc02-11e8-873d-5dbc9bd496e7.png)

- jv
![screenshot from 2018-09-19 11-56-03](https://user-images.githubusercontent.com/20714542/45726575-1e0ea480-bc03-11e8-88d0-749f35156712.png)

![screenshot from 2018-09-19 11-55-46](https://user-images.githubusercontent.com/20714542/45726576-1ea73b00-bc03-11e8-98e1-d9fb83079acc.png)

**Comment by fkazemian** [comment]
- [x] - py_source/lake_model.rst
- [x] - py_source/linear_algebra.rst
- [x] - py_source/linear_models.rst
- [x] - py_source/lln_clt.rst
- [x] - py_source/lqcontrol.rst
- [x] - py_source/lqramsey.rst
- [x] - py_source/lu_tricks.rst
- [x] - py_source/lucas_model.rst
- [x] - py_source/markov_asset.rst
- [x] - py_source/markov_perf.rst
- [x] - py_source/matplotlib.rst
- [x] - py_source/matsuyama.rst
- [x] - py_source/mccall_model.rst
- [x] - py_source/mccall_model_with_separation.rst
- [x] - py_source/mle.rst
- [x] - py_source/multiplicative_functionals.rst

- lake_model
![screenshot from 2018-09-19 12-20-08](https://user-images.githubusercontent.com/20714542/45727305-7b582500-bc06-11e8-96ed-86a6ac667e03.png)

- mle
![screenshot from 2018-09-19 12-53-42](https://user-images.githubusercontent.com/20714542/45728411-3c789e00-bc0b-11e8-9690-e463353f5262.png)

![screenshot from 2018-09-19 12-53-25](https://user-images.githubusercontent.com/20714542/45728412-3c789e00-bc0b-11e8-857a-9ec7b56412c1.png)

![screenshot from 2018-09-19 12-53-09](https://user-images.githubusercontent.com/20714542/45728413-3d113480-bc0b-11e8-997e-949095977762.png)

**Comment by fkazemian** [comment]
- [x] - py_source/numba.rst
- [x] - py_source/numpy.rst
- [x] - py_source/odu.rst
- [x] - py_source/ols.rst
- [x] - py_source/oop_intro.rst
- [x] - py_source/opt_tax_recur.rst
- [x] - py_source/optgrowth.rst
- [x] - py_source/orth_proj.rst
- [x] - py_source/pandas.rst
- [x] - py_source/pandas_panel.rst
- [x] - py_source/perm_income.rst
- [x] - py_source/perm_income_cons.rst
- [x] - py_source/python_advanced_features.rst
- [x] - py_source/python_by_example.rst
- [x] - py_source/python_essentials.rst
- [x] - py_source/python_oop.rst

- odu
![screenshot from 2018-09-19 13-14-44](https://user-images.githubusercontent.com/20714542/45729065-228c8a80-bc0e-11e8-942a-caad0853b6ec.png)

![screenshot from 2018-09-19 13-14-27](https://user-images.githubusercontent.com/20714542/45729066-228c8a80-bc0e-11e8-81be-1a03741a6f0f.png)

![screenshot from 2018-09-19 13-14-10](https://user-images.githubusercontent.com/20714542/45729067-228c8a80-bc0e-11e8-9025-1acfef88a01b.png)

- ols
![screenshot from 2018-09-19 13-16-29](https://user-images.githubusercontent.com/20714542/45729178-b52d2980-bc0e-11e8-9ea3-87d08f5207b7.png)

![screenshot from 2018-09-19 13-17-06](https://user-images.githubusercontent.com/20714542/45729181-b78f8380-bc0e-11e8-8a70-73ee653543a3.png)

![screenshot from 2018-09-19 13-18-41](https://user-images.githubusercontent.com/20714542/45729184-b9f1dd80-bc0e-11e8-9b48-847d91943c08.png)

- opt_tax_recur
![screenshot from 2018-09-19 13-27-29](https://user-images.githubusercontent.com/20714542/45729462-e9edb080-bc0f-11e8-9949-231fa3d50661.png)

- pandas
![screenshot from 2018-09-19 13-32-44](https://user-images.githubusercontent.com/20714542/45729613-992a8780-bc10-11e8-9de5-099a77f1eee2.png)

- python_advanced_features
![screenshot from 2018-09-19 13-48-55](https://user-images.githubusercontent.com/20714542/45730178-a301ba00-bc13-11e8-818e-4f058b8e5183.png)

![screenshot from 2018-09-19 13-49-19](https://user-images.githubusercontent.com/20714542/45730181-a4cb7d80-bc13-11e8-869e-384f0d4e4ca6.png)

![screenshot from 2018-09-19 13-53-23](https://user-images.githubusercontent.com/20714542/45730200-b745b700-bc13-11e8-8577-41a93a76b767.png)

- python_essentials
![screenshot from 2018-09-19 13-59-51](https://user-images.githubusercontent.com/20714542/45730345-78fcc780-bc14-11e8-8d33-1e3bfccd4944.png)

**Comment by fkazemian** [comment]
- [x] - py_source/rational_expectations.rst
- [x] - py_source/rob_markov_perf.rst
- [x] - py_source/robustness.rst
- [x] - py_source/samuelson.rst
- [x] - py_source/schelling.rst
- [x] - py_source/sci_libs.rst
- [x] - py_source/scipy.rst
- [x] - py_source/short_path.rst
- [x] - py_source/smoothing.rst
- [x] - py_source/stationary_densities.rst
- [x] - py_source/uncertainty_traps.rst
- [x] - py_source/wald_friedman.rst
- [x] - py_source/writing_good_code.rst

- robustness
![screenshot from 2018-09-19 14-09-23](https://user-images.githubusercontent.com/20714542/45730565-cded0d80-bc15-11e8-9bea-ac2492440ecc.png)

- samuelson
twice output
![screenshot from 2018-09-19 14-12-20](https://user-images.githubusercontent.com/20714542/45730697-85821f80-bc16-11e8-8a4b-537434b9f015.png)

![screenshot from 2018-09-19 14-14-34](https://user-images.githubusercontent.com/20714542/45730698-87e47980-bc16-11e8-8b49-0886b9dd6094.png)

- sci_libs
![screenshot from 2018-09-19 14-17-23](https://user-images.githubusercontent.com/20714542/45730779-1f49cc80-bc17-11e8-97cd-c83f64071bc4.png)
and after that, we can see lack of the output for all blocks
![screenshot from 2018-09-19 14-18-52](https://user-images.githubusercontent.com/20714542/45730837-6c2da300-bc17-11e8-9426-f6e2778b439b.png)

**Comment by fkazemian** [comment]
Hi @mmcky and @natashawatkins , I went through the lectures and looked for warnings, lectures with warnings in output are: 
py_source/additive_functionals.rst
py_source/amss.rst
py_source/amss2.rst
py_source/amss3.rst
py_source/arma.rst
py_source/chang_credible.rst
py_source/chang_ramsey.rst
py_source/discrete_dp.rst
py_source/estspec.rst
py_source/finite_markov.rst
py_source/hist_dep_policies.rst
py_source/jv.rst
py_source/lake_model.rst
py_source/mle.rst
py_source/odu.rst
py_source/ols.rst
py_source/opt_tax_recur.rst
py_source/pandas.rst
py_source/python_advanced_features.rst
py_source/python_essentials.rst
py_source/robustness.rst
py_source/samuelson.rst
py_source/sci_libs.rst

we can see the screenshot of warnings above.

**Comment by mmcky** [comment]
@natashawatkins and @fkazemian can I leave this issue with you to fix? Let me know if there are issues that can't be resolved within the source code and I will look at a way to fix in jupinx etc.

**Comment by natashawatkins** [comment]
@mmcky I've fixed one of the errors in mle in #60. Another one I don't think I can fix because it's in the `statsmodels` package, and the other is left in on purpose (the surrounding text looks at the warning message)

**Comment by mmcky** [comment]
I think the majority of these are now fixed (i.e. `ols` is not showing any errors)
I propose to close this and open issues if any additional ones crop up.
